src/functional.jl

```diff
@@ -0,0 +1,111 @@
+# finite-dimensional vector spaces — we attempt to support most operations on
+# arrays except the linear algebra related ones; the infinite-dimensional nature
+# makes subtyping AbstractArray a little dangerous, which explains why
+# functionality is reimplemented instead of subtyping AbstractArray
+
+# size   – not implemented since an infinite-dimensional tuple would be strange
+# length – one or zero
+# endof  – same as length
+length(u::Nullable) = u.isnull ? 0 : 1
+endof(u::Nullable)  = length(u)
+
+# indexing is either without index, or with 1 as index
+# generalized linear indexing is not supported
+# setindex! not supported because Nullable is immutable
+linearindexing{T}(::Nullable{T}) = LinearFast()
+function getindex(u::Nullable)
+    @boundscheck u.isnull && throw(NullException())
+    u.value
+end
+function getindex(u::Nullable, i::Integer)
+    @boundscheck u.isnull | (i ≠ one(i)) && throw(BoundsError(i, u))
+    u.value
+end
+
+# iteration protocol
+start(u::Nullable) = 1
+next(u::Nullable, i::Integer) = u.value, 0
+done(u::Nullable, i::Integer) = u.isnull || i == 0
+
+# next we have reimplementations of some higher-order functions
+filter{T}(p, u::Nullable{T}) = u.isnull ? u : p(u.value) ? u : Nullable{T}()
+
+# warning: not type-stable
+map{T}(f, u::Nullable{T}) = u.isnull ? Nullable{Union{}}() : Nullable(f(u.value))
+
+# multi-argument map doesn't broadcast, so not very useful, but no harm having
+# it...
+function map(f, us::Nullable...)
+    if all(isnull, us)
+        Nullable()
+    elseif !any(isnull, us)
+        Nullable(map(f, map(getindex, us)...))
+    else
+        throw(DimensionMismatch("expected all null or all nonnull"))
+    end
+end
+
+# foldr and foldl are quite useful to express "do something if not null, else"
+# these
+
+# being infinite-dimensional, nullables are generally incompatible with
+# broadcast with arrays — it is probably not worth supporting the rare case
+# where an array has length 0 or 1
+
+# so we reimplement broadcast, which has the same semantics but very different
+# implementation. This implementation is in fact much simpler than that for
+# arrays. Length-1 (non-nulls) are flexible and can broadcast to length-0
+# (nulls), but the other way does not work. Numbers are zero-dimensional and can
+# broadcast to infinite-dimensional nullables, but the other direction is not
+# supported.
+
+# Base's broadcast has a very loose signature so we can easily make it more
+# specific. Broadcast on numbers is still not supported. FIXME: remove generated
+# functions where unnecessary
+
+# some specialized functions
+broadcast{T}(f, u::Nullable{T}) =
+    u.isnull ? Nullable{promote_op(f, T)}() : Nullable{promote_op(f, T)}(f(u.value))
+
+@generated function broadcast(f, u::Union{Nullable,Number}, v::Union{Nullable,Number})
+    checkfor(s) = :($s.isnull && return Nullable{result}())
+    lifted = [T <: Nullable ? T.parameters[1] : T for T in (u, v)]
```

Comment:
`T.parameters` can be replaced with `eltype(T)`. Also, I would find it clearer to compute `U` and `V` separately instead of creating a vector and splatting it below.